compiler/luci/pass/src/SubstituteExpandDimsToReshapePass.test.cpp

```diff
@@ -0,0 +1,201 @@
+
+    _output->from(_expand_dims);
+  }
+
+protected:
+  luci::CircleExpandDims *_expand_dims = nullptr;
+};
+
+class SubstituteExpandDimsToReshapeTest : public ::testing::Test
+{
+public:
+  SubstituteExpandDimsToReshapeTest() = default;
+
+  void run_pass(void)
+  {
+    while (_shapeinf.run(_graph.g()) || _pass.run(_graph.g()))
+      ;
+  }
+
+protected:
+  SubstituteExpandDimsToReshapeGraph _graph;
+  luci::SubstituteExpandDimsToReshapePass _pass;
+  luci::CircleShapeInferencePass _shapeinf;
+};
+
+} // namespace
+
+TEST(SubstituteExpandDimsToReshapePassTest, name)
+{
+  luci::SubstituteExpandDimsToReshapePass pass;
+  auto const name = pass.name();
+  ASSERT_NE(nullptr, name);
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, simple_with_expand_dims_1)
+{
+  _graph.init({2, 16}, {2, 1, 16}, 1);
+
+  run_pass();
+
+  auto reshape = dynamic_cast<luci::CircleReshape *>(_graph.output()->from());
+  auto expand_dims = dynamic_cast<luci::CircleExpandDims *>(_graph.output()->from());
+  ASSERT_NE(nullptr, reshape);
+  ASSERT_EQ(nullptr, expand_dims);
+  auto reshape_shape = loco::must_cast<luci::CircleConst *>(reshape->shape());
+  ASSERT_EQ(3, reshape_shape->size<loco::DataType::S32>());
+  ASSERT_EQ(2, reshape_shape->at<loco::DataType::S32>(0));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(1));
+  ASSERT_EQ(16, reshape_shape->at<loco::DataType::S32>(2));
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, simple_with_expand_dims_M1)
+{
+  _graph.init({2, 3, 4}, {2, 3, 4, 1}, -1);
+
+  run_pass();
+
+  auto reshape = dynamic_cast<luci::CircleReshape *>(_graph.output()->from());
+  auto expand_dims = dynamic_cast<luci::CircleExpandDims *>(_graph.output()->from());
+  ASSERT_NE(nullptr, reshape);
+  ASSERT_EQ(nullptr, expand_dims);
+  auto reshape_shape = loco::must_cast<luci::CircleConst *>(reshape->shape());
+  ASSERT_EQ(4, reshape_shape->size<loco::DataType::S32>());
+  ASSERT_EQ(2, reshape_shape->at<loco::DataType::S32>(0));
+  ASSERT_EQ(3, reshape_shape->at<loco::DataType::S32>(1));
+  ASSERT_EQ(4, reshape_shape->at<loco::DataType::S32>(2));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(3));
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, simple_with_expand_dims_2)
+{
+  _graph.init({16, 3, 1}, {16, 3, 1, 1}, 2);
+
+  run_pass();
+
+  auto reshape = dynamic_cast<luci::CircleReshape *>(_graph.output()->from());
+  auto expand_dims = dynamic_cast<luci::CircleExpandDims *>(_graph.output()->from());
+  ASSERT_NE(nullptr, reshape);
+  ASSERT_EQ(nullptr, expand_dims);
+  auto reshape_shape = loco::must_cast<luci::CircleConst *>(reshape->shape());
+  ASSERT_EQ(4, reshape_shape->size<loco::DataType::S32>());
+  ASSERT_EQ(16, reshape_shape->at<loco::DataType::S32>(0));
+  ASSERT_EQ(3, reshape_shape->at<loco::DataType::S32>(1));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(2));
+  ASSERT_EQ(1, reshape_shape->at<loco::DataType::S32>(3));
+}
+
+TEST_F(SubstituteExpandDimsToReshapeTest, nothing_to_expand_dims)
```

Comment:
Could you change this to a negative test and add more negative tests so that the number of negative tests == non-negative tests?

Reply:
I added some. :-D